File: backend/pkg/handlers/web/cpuIssue.go
package web

import (
	"encoding/json"
	"log"

	. "openreplay/backend/pkg/messages"
	"openreplay/backend/pkg/messages/performance"
)

/*
	Handler name: CpuIssue
	Input events: PerformanceTrack,
				  SetPageLocation
	Output event: IssueEvent
*/

const CpuThreshold = 70 // % out of 100
const CpuMinDurationTrigger = 6 * 1000

type CpuIssueDetector struct {
	startTimestamp uint64
	startMessageID uint64
	lastTimestamp  uint64
	maxRate        uint64
	contextString  string
}

func (f *CpuIssueDetector) createPayload() string {
	p, err := json.Marshal(struct {
		Duration uint64
		Rate     uint64
	}{f.duration(), f.maxRate})
	if err != nil {
		log.Printf("can't marshal CpuIssue payload to json: %s", err)
	}
	return string(p)
}

func (f *CpuIssueDetector) duration() uint64 {
	return f.lastTimestamp - f.startTimestamp
}

func (f *CpuIssueDetector) reset() {
	f.startTimestamp = 0
	f.startMessageID = 0
	f.maxRate = 0
}

func (f *CpuIssueDetector) Build() Message {
	defer f.reset()
	if f.startTimestamp == 0 || f.duration() < CpuMinDurationTrigger {
		return nil
	}
	return &IssueEvent{
		Type:          "cpu",
		Timestamp:     f.startTimestamp,
		MessageID:     f.startMessageID,
		ContextString: f.contextString,
		Payload:       f.createPayload(),
	}
}

func (f *CpuIssueDetector) Handle(message Message, timestamp uint64) Message {
	switch msg := message.(type) {
	case *PerformanceTrack:
		// Ignore if it's a wrong message order
		if timestamp < f.lastTimestamp {
			return nil
		}
		f.lastTimestamp = timestamp
		cpuRate := performance.CPURate(msg.Ticks, performance.TimeDiff(timestamp, f.lastTimestamp))
		// Build event if cpu issue have gone
		if msg.Frames == -1 || msg.Ticks == -1 || cpuRate < CpuThreshold {
			return f.Build()
		}
		// Update values
		if f.startTimestamp == 0 {
			f.startTimestamp = timestamp
			f.startMessageID = message.MsgID()
		}
		if f.maxRate < cpuRate {
			f.maxRate = cpuRate
		}
	case *SetPageLocation:
		f.contextString = msg.URL
	}
	return nil
}

File: tracker/tracker/src/main/detectors/cpuIssue.ts
import Message, {
  Type,
  PerformanceTrack,
  SetPageLocation,
} from '../../common/messages.gen.js'
import type { Detector, ReportIssue } from './types.js'
import { cpuRate, timeDiff } from './performance.js'

/**
 * Port of backend/pkg/handlers/web/cpuIssue.go.
 *
 * Tracks stretches where CPU load (derived from PerformanceTrack ticks) stays
 * above CPU_THRESHOLD for at least CPU_MIN_DURATION_TRIGGER ms.
 *
 * NOTE: the Go version overwrites lastTimestamp *before* computing the interval,
 * so its dt is always 0 (cpuRate always 0 -> effectively disabled). We compute
 * dt against the previous sample so the rate is meaningful for debugging.
 */

const CPU_THRESHOLD = 70 // % out of 100
const CPU_MIN_DURATION_TRIGGER = 6 * 1000

export default class CpuIssueDetector implements Detector {
  private startTimestamp = 0
  private lastTimestamp = 0
  private maxRate = 0
  private contextString = ''

  constructor(private readonly report: ReportIssue) {}

  private duration(): number {
    return this.lastTimestamp - this.startTimestamp
  }

  private reset(): void {
    this.startTimestamp = 0
    this.maxRate = 0
  }

  private build(): void {
    const start = this.startTimestamp
    const rate = this.maxRate
    const duration = this.duration()
    // matches Go `defer f.reset()`
    this.reset()
    if (start === 0 || duration < CPU_MIN_DURATION_TRIGGER) {
      return
    }
    this.report({
      type: 'cpu',
      contextString: this.contextString,
      payload: JSON.stringify({ Duration: duration, Rate: rate }),
      timestamp: start,
    })
  }

  handle(message: Message, timestamp: number): void {
    const type = message[0]
    if (type === Type.SetPageLocation) {
      this.contextString = (message as SetPageLocation)[1]
      return
    }
    if (type !== Type.PerformanceTrack) {
      return
    }
    const msg = message as PerformanceTrack
    const frames = msg[1]
    const ticks = msg[2]
    if (timestamp < this.lastTimestamp) {
      return
    }
    const prevTimestamp = this.lastTimestamp
    this.lastTimestamp = timestamp
    if (prevTimestamp === 0) {
      // first sample: no interval to measure the rate against yet
      return
    }
    const rate = cpuRate(ticks, timeDiff(timestamp, prevTimestamp))
    if (frames === -1 || ticks === -1 || rate < CPU_THRESHOLD) {
      this.build()
      return
    }
    if (this.startTimestamp === 0) {
      this.startTimestamp = timestamp
    }
    if (this.maxRate < rate) {
      this.maxRate = rate
    }
  }
}
